Hydra12 urlshortner vzeten 03 bjs2 97369 bjs2 97371 bjs2 97366

vzeten opened the pull request:
## URL Shortener Service - Implementation

### Описание
Реализован микросервис для сокращения URL с использованием Spring Boot, PostgreSQL и Redis.

### Выполненные задачи
- [x] BJS2-97370 — БД (Liquibase, таблицы url/hash, sequence)
- [x] BJS2-97367 — Base62Encoder для кодирования хэшей
- [x] BJS2-97372 — HashRepository (JdbcTemplate)
- [x] BJS2-97373 — HashGenerator (асинхронная генерация каждые 5 мин)
- [x] BJS2-97374 — HashCache (thread-safe очередь)
- [x] BJS2-97371 — POST /shorten (создание короткой ссылки)
- [x] BJS2-97369 — GET /{hash} (редирект 302)
- [x] BJS2-97368 — UrlExceptionHandler (503/500/404/400)
- [x] BJS2-97366 — CleanerScheduler (чистка старых URL в 3:00)

### Технические детали
- **Генерация хэшей:** Base62 кодирование (минимум 6 символов)
- **Кеширование:** Redis + in-memory Queue
- **Scheduler:** 
  - Генерация хэшей каждые 5 минут (async)
  - Удаление URL старше 1 года (ежедневно в 3:00)
- **Обработка ошибок:** Централизованный @RestControllerAdvice
- **Логирование:** SLF4J через @Slf4j

### Дополнительно
- Unit-тесты (Base62EncoderTest, UrlServiceTest)
- Логирование всех ключевых операций
- README.md с полной документацией

Commit:
Сделал задачи
- [x] Создать структуру БД - BJS2-97370
- [x] Генерация списка хэшей - BJS2-97367
- [x] Работа с базой(HashRepository) - BJS2-97372

Commit:
Поправил замечания

Commit:
Что сделано:
ErrorResponse.java — DTO record с полями (timestamp, status, error, message, path)

NoFreeHashesException.java — кастомное исключение

UrlExceptionHandler.java — @RestControllerAdvice с двумя методами (503 для NoFreeHashesException, 500 для Exception)

HashGenerator.java — добавили @Scheduled(cron = "0 */5 * * * ?")

ServiceTemplateApplication.java — добавили @EnableScheduling

application.yaml — добавили spring.task.scheduling.pool.size: 10

Commit:
## URL Shortener Service - Implementation

### Описание
Реализован микросервис для сокращения URL с использованием Spring Boot, PostgreSQL и Redis.

### Выполненные задачи
- [x] BJS2-97370 — БД (Liquibase, таблицы url/hash, sequence)
- [x] BJS2-97367 — Base62Encoder для кодирования хэшей
- [x] BJS2-97372 — HashRepository (JdbcTemplate)
- [x] BJS2-97373 — HashGenerator (асинхронная генерация каждые 5 мин)
- [x] BJS2-97374 — HashCache (thread-safe очередь)
- [x] BJS2-97371 — POST /shorten (создание короткой ссылки)
- [x] BJS2-97369 — GET /{hash} (редирект 302)
- [x] BJS2-97368 — UrlExceptionHandler (503/500/404/400)
- [x] BJS2-97366 — CleanerScheduler (чистка старых URL в 3:00)

### Технические детали
- **Генерация хэшей:** Base62 кодирование (минимум 6 символов)
- **Кеширование:** Redis + in-memory Queue
- **Scheduler:**
  - Генерация хэшей каждые 5 минут (async)
  - Удаление URL старше 1 года (ежедневно в 3:00)
- **Обработка ошибок:** Централизованный @RestControllerAdvice
- **Логирование:** SLF4J через @Slf4j

### Дополнительно
- Unit-тесты (Base62EncoderTest, UrlServiceTest)
- Логирование всех ключевых операций
- README.md с полной документацией

AlexeyAntipin commented on src/main/java/faang/school/urlshortenerservice/scheduler/CleanerScheduler.java:
```diff
@@ -0,0 +1,30 @@
+    private final HashRepository hashRepository;
+
+    @Transactional
+    @Scheduled(cron = "0 0 3 * * ?")  // ← какое расписание? (например, раз в день)
```
Лучше вынести cron в application.yaml

AlexeyAntipin commented on src/main/java/faang/school/urlshortenerservice/service/UrlService.java:
```diff
@@ -0,0 +1,58 @@
+        urlRepository.save(hash, longUrl);
+        urlCacheRepository.save(hash, longUrl);
+
+        String shortUrl = baseUrl + "/" + hash;
```
Лучше не конкатенировать строки и использовать String.format

AlexeyAntipin commented on src/main/java/faang/school/urlshortenerservice/generator/HashCache.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    private void refillCacheIfNeeded() {
+        if (cache.size() >= minCacheSize) {
```
refill должен происходить в отдельном потоке

AlexeyAntipin commented on src/main/java/faang/school/urlshortenerservice/repository/UrlCacheRepository.java:
```diff
@@ -0,0 +1,22 @@
+@RequiredArgsConstructor
+public class UrlCacheRepository {
+
+    private final StringRedisTemplate redisTemplate;  // ← Spring bean для Redis
```
Ты проверял, что все работает? Мне почему-то казалось, что редис из коробки, без дополнительных конфигураций не умет конфигурироваться

vzeten replied:
Проверял, работает

AlexeyAntipin reviewed (commented):
Я бы еще добавил наполнения кэша при запуске приложения

Commit:
1. HashCache.java:
   - Добавлен async refill через Executor
   - Защита от race condition (AtomicBoolean)
   - Инициализация кэша при старте (@PostConstruct)

2. UrlService.java:
   - Заменил конкатенацию на String.format

3. application.yml:
   - Вынес cron в конфиг (scheduler.cleaner.cron)

4. CleanerScheduler.java:
   - Cron читается из application.yml
   - Добавлена проверка пустого списка перед save

AlexeyAntipin commented:
Конфликты осталось только порешать

vzeten committed:
Merge branch 'hydra12-urlshortner-vzeten' into hydra12-urlshortner-vzeten-03-BJS2-97369-BJS2-97371-BJS2-97366
